File: 337.rob.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, int> memT;
    unordered_map<TreeNode*, int> memF;

    int cnt(TreeNode* root, bool parent) {
        if (root == NULL) return 0;
        if (parent && memT[root]) return memT[root];
        if (!parent && memF[root]) return memF[root];

        int child = 0;
        if (parent) {
            child = cnt(root->left, false) + cnt(root->right, false);
            memT[root] = child;
        } else {
            child = root->val + cnt(root->left, true) + cnt(root->right, true);
            child = max(child, cnt(root->left, false) + cnt(root->right, false));
            memF[root] = child;
        }
        return child;
    }

    int rob(TreeNode* root) {
        if (root == NULL) return 0;
        return max(cnt(root->left, false)+cnt(root->right, false), root->val + cnt(root->left, true) + cnt(root->right, true));
    }
};

File: 337.rob.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int rob(TreeNode root) {
        int[] res = robNode(root);
        return Math.max(res[0], res[1]);
    }

    public int[] robNode(TreeNode node) {
        if (node == null) return new int[]{0,0};
        int[] left = robNode(node.left);
        int[] right = robNode(node.right);

        int r1 = node.val + left[0] + right[0];
        int r2 = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);

        return new int[]{r2, r1};
    }
}
